ion_hash_test.go

```diff
@@ -0,0 +1,595 @@
+
+					for reader.Next() {
+						annotations := reader.Annotations()
+
+						if len(annotations) > 0 {
+							if annotations[0] == "update" {
+								updateBytes := readSexpAndAppendToList(t, reader)
+								identityUpdateList = append(identityUpdateList, updateBytes)
+
+							} else if annotations[0] == "digest" {
+								digestBytes := readSexpAndAppendToList(t, reader)
+								identityDigestList = append(identityDigestList, digestBytes)
+							} else if annotations[0] == "final_digest" {
+								digestBytes := readSexpAndAppendToList(t, reader)
+								identityFinalDigest = append(identityFinalDigest, digestBytes)
+							}
+						}
+					}
+					err = reader.StepOut()
+					assert.NoError(t, err, "Something went wrong executing reader.StepOut()")
+				} else if fieldName == "md5" {
+					err = reader.StepIn()
+					assert.NoError(t, err, "Something went wrong executing reader.StepIn()")
+
+					for reader.Next() {
+						annotations := reader.Annotations()
+
+						if len(annotations) > 0 {
+							if annotations[0] == "update" {
+								updateBytes := readSexpAndAppendToList(t, reader)
+								md5UpdateList = append(md5UpdateList, updateBytes)
+
+							} else if annotations[0] == "digest" {
+								digestBytes := readSexpAndAppendToList(t, reader)
+								md5DigestList = append(md5DigestList, digestBytes)
+							}
+						}
+					}
+					err = reader.StepOut()
+					assert.NoError(t, err, "Something went wrong executing reader.StepOut()")
+				}
+
+				expectedHashLog := hashLog{
+					identityUpdateList:  identityUpdateList,
+					identityDigestList:  identityDigestList,
+					identityFinalDigest: identityFinalDigest,
+					md5UpdateList:       md5UpdateList,
+					md5DigestList:       md5DigestList,
+				}
+
+				if hasherName != "identity" {
+					testName = testName + "." + hasherName
+				}
+
+				dataList = append(dataList, testObject{testName, testCase, expectedHashLog, newTestIonHasherProvider(hasherName)})
+			}
+			err = reader.StepOut()
+			assert.NoError(t, err, "Something went wrong executing reader.StepOut()")
+		}
+		err = reader.StepOut()
+		assert.NoError(t, err, "Something went wrong executing reader.StepOut()")
+	}
+	return dataList
+}
+
+type testObject struct {
+	hasherName      string
+	testCase        []byte
+	expectedHashLog hashLog
+	provider        *testIonHasherProvider
+}
+
+func writeToWriter(t *testing.T, reader ion.Reader, textWriter ion.Writer, binaryWriter ion.Writer) {
+	ionType := reader.Type()
+
+	if reader.Annotations() != nil {
+		err := textWriter.Annotations(reader.Annotations()...)
+		require.NoError(t, err)
+		err = binaryWriter.Annotations(reader.Annotations()...)
+		require.NoError(t, err)
+	}
+
+	switch ionType {
+	case ion.NullType:
+		if reader.FieldName() != "" {
+			err := textWriter.FieldName(reader.FieldName())
+			require.NoError(t, err)
+			err = binaryWriter.FieldName(reader.FieldName())
+			require.NoError(t, err)
+		}
```

Comment:
Looks like this FieldName logic applies to Null and all Scalar types.

How about we put this logic in a `if ionType == ion.NullType || ion.IsScalar(ionType)` block before the `switch ionType` block rather than repeating it for each of those types.

Reply:
FieldName applies to all types if they are in a struct already. I am trying to find a way so it doesn't get repeated

Reply:
Updated